### PR TITLE
[REVIEW] Allow writing parquet to python file like objects

### PR DESCRIPTION
**Pull request by benfred**
Add code that allows writing parquet files to python file-like objects. This
lets you write cudf dataframes as parquet to BytesIO or SocketIO objects,
without having to write out to the disk first.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by benfred** [comment]
We've noticed that the reading/writing parquet files is really fast with cudf, and this change lets spill to host memory using parquet, without hitting disk by writing to a python BytesIO object. This also lets you right directly to the network using SocketIO etc.

On a small test this conversion is almost twice as fast as converting to/from pandas:

```python
In [4]: def convert_to_from_pandas(gdf): 
   ...:     pdf = gdf.to_pandas() 
   ...:     cudf.from_pandas(pdf) 
   ...:                                                                                                       

In [5]: def convert_to_from_parquet(gdf): 
   ...:     buffer = io.BytesIO() 
   ...:     gdf.to_parquet(buffer) 
   ...:     cudf.read_parquet(buffer) 
   ...:                                                                                                       

In [6]: time convert_to_from_pandas(gdf)                                                                      
CPU times: user 2.04 s, sys: 572 ms, total: 2.62 s
Wall time: 2.62 s

In [7]: time convert_to_from_parquet(gdf)                                                                     
CPU times: user 753 ms, sys: 600 ms, total: 1.35 s
Wall time: 1.35 s

File: python/cudf/cudf/_lib/io/utils.pyx [hunk 1]
```diff
@@ -27,3 +31,35 @@ cdef source_info make_source_info(src) except*:
...
+        with gil:
+            buf.write(PyMemoryView_FromMemory(<char*>data, size, PyBUF_READ))
+
+    void flush() nogil:
+        with gil:
+            buf.flush()
+
+    size_t bytes_written() nogil:
+        with gil:
+            return buf.tell()
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/io/utils.pyx:65]: Any reason to mark these as `nogil` and then all of the functions need the gil so they use a context `with gil:`?

Comment (kkraus14) [reply python/cudf/cudf/_lib/io/utils.pyx:65]: Did you try just doing?
```
void flush() with gil:
    buf.flush()
```

This just feels no bueno that we could be thrashing releasing and reacquiring the GIL unnecessarily.

**Comment by kkraus14** [comment]
> We might be better off using the ‘HOST_BUFFER’ sink type, which writes to a `std::vector<char>`. We could expose this as an opaque type to Python without too much effort. I can add this if you and Olivier think it's the way to go.

I think this would likely be the highest performance and sets us up to do things more asynchronously moving forward. Additionally, we could likely wrap the returned buffer in a class that inherits from IOBase zero-copy to still make it a file-like object. What would be interesting about this approach as well is it gives us more fine grained control over synchronization to only synchronize if we're trying to use non-CUDA aware methods against it that don't respect stream ordering.


> I think there are some benefits to this IOBase implementation that outweighs the downsides of an extra memcpy though: this more closely matches the to_parquet implementation in pandas which supports writing to any file-like object, and this lets you do things like write directly to a python SocketIO or File.

Agreed. We should keep this to allow people to write to an arbitrary file-like object that they provide.

File: python/cudf/cudf/_lib/io/utils.pyx (hunk 1 above)
Comment (benfred) [reply python/cudf/cudf/_lib/io/utils.pyx]: I hadn't tried that - and it works great, changed in latest commit. Thanks!

**Comment by rjzamora** [comment]
This is great @benfred - Thank you for all the work here!

@kkraus14 @madsbk  - Note that the motivation for both this and #4870 (both by @benfred) is larger-than-memory shuffling on a single GPU.  I can imagine these features being used in combination with Mads’ [explicit comms](https://github.com/rapidsai/dask-cuda/blob/branch-0.14/dask_cuda/explicit_comms/comms.py) to dramatically reduce the size of the task graph needed for shuffling in dask_cudf ([discussion here](https://github.com/dask/dask/issues/6163) for those unfamiliar with the problem).   The algorithm is not likely to get upstream-dask/distributed support (since it is explicit-comms based, and resiliency would be a problem), but it could be useful in some cases. My idea goes something like this:

- The primary API would take in a `dask_cudf.DataFrame` (call it `a`) with `npartitions` partitions
- Since explicit communication requires that the input ddf be persisted in memory, we would “iterate” through `a` in chunks that can fit comfortably in global worker device memory.  For example, each iteration might correspond to `n` partitions each. For each iteration `i`, we simply define a new ddf (`b_i`), comprising a subset of `n` partitions.
- For iteration `i`, we perform an explicit shuffle (call it `_shuffle_to_pqcache`) on `b_i` (corresponding to a single “super” task).  The `_shuffle_to_pqcache` logic will split each partition into `npartitions` shards, and send each shard to the worker responsible for that shard (use simple modulus mapping).  After exchanging shards, each worker can create or append to a separate cached parquet file (in host memory) for each output partition that worker is responsible for.  The actual output of the `_shuffle_to_pqcache` task can be a simple boolean confirmation that the exchange and caching is done. 
- After all iterations of `_shuffle_to_pqcache` are completed, we would need some kind of follow-up task(s) (called something like `_worker_pqcache_to_ddf`) to convert the in-memory parquet files, on each worker, into the new output partitions.  I have not thought through this particular step much yet, so I apologize if I am failing to think of an obvious roadblock here.

@madsbk - Since you are most familiar with the nuts and bolts of explicit communication, do you think this general approach has any promise?

**Comment by benfred** [comment]
> I think this would likely be the highest performance and sets us up to do things more asynchronously moving forward. Additionally, we could likely wrap the returned buffer in a class that inherits from IOBase zero-copy to still make it a file-like object. What would be interesting about this approach as well is it gives us more fine grained control over synchronization to only synchronize if we're trying to use non-CUDA aware methods against it that don't respect stream ordering.

That sounds like a good plan to me - I'll submit a new PR for this later today

**Pull request merged**